plugins/module_utils/module_helper.py

```diff
@@ -282,40 +282,42 @@ def custom_formats(self):
             result[param] = ArgFormat(param, **fmt_spec)
         return result
 
-    def _calculate_args(self, extra_params=None, params=None):
+    def _calculate_args(self, params=None):
         def add_arg_formatted_param(_cmd_args, arg_format, _value):
             args = list(arg_format.to_text(_value))
             return _cmd_args + args
 
         def find_format(_param):
             return self.custom_formats.get(_param, self.module_formats.get(_param))
 
-        extra_params = extra_params or dict()
         cmd_args = list([self.command]) if isinstance(self.command, str) else list(self.command)
         cmd_args[0] = self.module.get_bin_path(cmd_args[0])
         param_list = params if params else self.module.params.keys()
 
         for param in param_list:
-            if param in self.module.argument_spec:
-                if param not in self.module.params:
-                    continue
-                fmt = find_format(param)
-                value = self.module.params[param]
+            if isinstance(param, dict):
+                if len(param.keys()) != 1:
+                    raise ModuleHelperException("run_command parameter as a dict must "
+                                                "contain only one key: {0}".format(param))
+                _param = list(param.keys())[0]
+                fmt = find_format(_param)
+                value = param[_param]
+            elif isinstance(param, str):
+                if param in self.module.argument_spec:
+                    fmt = find_format(param)
+                    value = self.module.params[param]
+
             else:
-                if param not in extra_params:
-                    continue
-                fmt = find_format(param)
-                value = extra_params[param]
-            self.cmd_args = cmd_args
+                raise ModuleHelperException("run_command parameter must be either a str or a dict: {0}".format(param))
             cmd_args = add_arg_formatted_param(cmd_args, fmt, value)
 
         return cmd_args
 
     def process_command_output(self, rc, out, err):
         return rc, out, err
 
-    def run_command(self, extra_params=None, params=None, *args, **kwargs):
-        self.vars['cmd_args'] = self._calculate_args(extra_params, params)
+    def run_command(self, params=None, *args, **kwargs):
```

Comment:
This is a breaking change and needs to be deprecated first, before it can be removed in a future major release.

Reply:
I  was afraid you were going to say that. I don't suppose the fact that no one else is using ModuleHelper helps me here, does it?

Reply:
Not really: the `ModuleHelper` is part of the public interface of the collection, and other (public or private) collections (or custom modules) could use it. While I agree that this isn't very likely, it doesn't help with that this is a breaking change :)

Reply:
Fair enough. Given that, I am reverting the removal of the parameter (and associated behaviour), so that we prevent the breaking change for now, but I am keeping the new future of handling `dict` parameters.

Reply:
Just for the sake of learning: if I could jump into a DeLorean and go back in time a couple of months, what would be the right/best option to have `ModuleHelper` **not** be part of the public interface of the collection, and yet be available to mode than one module? Right now `xfconf` is the only one using it, but I maintain `django_manage` as well and I have been planning on refactoring it to use `ModuleHelper`.

In my head the idea was to use MH myself for a while, find the more obvious improvements and then at some point make it public, start promoting it to other module developers.

Reply:
@russoz everything in module_utils is part of the public interface, though it's less obvious than other parts. I guess the only way would have been to explicitly declare it as private and put an underscore at the beginning of the filename.

Reply:
More you live, more you learn. Thanks for clarifying, will keep it mind for the future.